### PR TITLE
Document and publish pluggable allocators header file.

### PR DESCRIPTION
**Pull request by anakrish**
allocators.h will be installed to openenclave/include/advanced/allocator.h

Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

### CODE REVIEW

**Comment by anakrish** [comment]
bors try

File: include/openenclave/advanced/allocator.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ * @param[in] heap_end_address The ending address of the enclave heap.
+ */
+void oe_allocator_init(void* heap_start_address, void* heap_end_address);
+
+/**
+ * Callback for cleaning up the allocator.
+ *
+ * This function will be called by oecore at an appropriate time during enclave
+ * termination. atexit functions will be executed prior to oe_allocator_cleanup.
+ * oe_allocator_cleanup may not register functions with atexit.
```
Comment (CodeMonkeyLeet) [inline include/openenclave/advanced/allocator.h:45]: Similarly, consider simplifying:
```suggestion
 * This function will be called by oecore during enclave termination.
 * atexit functions will be executed prior to oe_allocator_cleanup.
 * oe_allocator_cleanup may not register functions with atexit.
```

File: include/openenclave/advanced/allocator.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+ * @endcond
+ */
+
+/**
+ * Callback for initializing the allocator.
+ *
+ * This function will be called by oecore at an appropriate time during enclave
+ * initialization. The allocator is expected to perform any global
+ * initialization in this function. This function is called prior to any global
+ * initialization.
```
Comment (CodeMonkeyLeet) [inline include/openenclave/advanced/allocator.h:32]: Consider clarifying the subject of the global initialization in either case:
``` suggestion 
 * The allocator is expected to perform any global initialization needed by the
 * allocator in this function. This function is called prior to other enclave
 * initialization methods, which can depend on the allocator being ready for
 * use.
```

File: include/openenclave/advanced/allocator.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+void oe_allocator_thread_cleanup(void);
+
+/**
+ * Callback to allocate memory.
+ *
+ * This function will be called by oecore to implement malloc.
+ * oe_allocator_malloc must provide the same semantics as malloc.
+ * See: https://en.cppreference.com/w/c/memory/malloc
+ *
+ * @param[size] Number of bytes to allocate.
```
Comment (dthaler) [inline include/openenclave/advanced/allocator.h:80]:
```suggestion
 * @param[in] size Number of bytes to allocate.
```
incorrect doxygen syntax in the file.   Fix here and below.

Comment (anakrish) [reply include/openenclave/advanced/allocator.h]: Good catch. Fixed.

File: include/openenclave/advanced/allocator.h (hunk 2 above)
Comment (anakrish) [reply include/openenclave/advanced/allocator.h:32]: Thanks!

File: include/openenclave/advanced/allocator.h (hunk 1 above)
Comment (anakrish) [reply include/openenclave/advanced/allocator.h:45]: Done.

**Comment by anakrish** [comment]
> LGTM, and I found the doxygen comments really helpful!
> 
> Is this the last bit of work for the pluggable allocators? If so, also consider updating the CHANGELOG.md to describe the added feature.

I added an entry in the CHANGELOG.
There will be one more PR to add a sample.

**Comment by anakrish** [comment]
bors try

**Pull request merged**